app/dts/bindings/behaviors/zmk,behavior-macro.yaml

```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2020 The ZMK Contributors
+# SPDX-License-Identifier: MIT
+
+description: Macro
+
+compatible: "zmk,behavior-macro"
+
+include: zero_param.yaml
+
+properties:
```

Comment:
That's a great idea. We can actually implement `&down` and `&up` and `&wait` independently from this PR.

I agree we should remove the delay that now exists between separate keypresses so people can set up their own `&waits`.

Reply:
More macro ideas:

* repeat boolean. If it's on, the macro repeats forever while it's held down.

Reply:
If you do remove the delay between key presses, note that `&kp` will probably still need a delay between press and release. Maybe there should still be a configurable duration for `&kp` and then anything else you do with wait commands?

Reply:
After thinking about this some more, having a special case to put a delay between press and release for just `&kp` seems weird. What if instead we added a duration value to each item, and then made macro execution work like this?

1. Press item
2. Wait item's duration
3. Release item
4. Move to next item and immediately return to step 1

Then for a delay between items you could just use `&none` with the duration you want to delay by.

I need to read up more on how Zephyr handles lists of DT things, so this might not be possible or there might be a better way to do it, but maybe we could add a special binding which the macro code looks for, and instead of pressing/releasing it, it uses its parameter to determine how long the following item should be held? If an item is not preceded by a duration, use something like the `sleep` property as a default (and default that property to 0).

```
default-duration = <10>;
bindings =
    <&for 0 &down LALT>,
    <&kp TAB>,
    <&for 30 &none>.
    <&kp TAB>,
    <&for 0 &up LALT>,
    <&kp LGUI(LEFT)>;
```

Alternatively (and much more awkwardly), we could have parallel arrays of bindings and durations:

```
bindings = <&down LALT>, <kp TAB>, <&none>, <&kp TAB>, <&up LALT>, <&kp LGUI(LEFT)>;
duration = <0>,          <10>,     <30>,    <10>,      <0>,        <10>;
```

Reply:
Seems like a default duration and `<&up LC> <&sleep X> <&down KC>` would be much simpler:
```
keypress-duration = <10>;
bindings =
    <&press LALT>,
    <&kp TAB>,
    <&sleep 30>.
    <&kp TAB>,
    <&release LALT>,
    <&kp LGUI(LEFT)>;
``` 

(`&press` and `&release` should not invoke any sleeps)

Reply:
If I'm not mistaken, I think with the current iteration, you can get quite a bit of this functionality by nesting the different macros. You can get the different wait times and press/release keys with just the current implementation of KP. It would be nice to simplify it with the other features, but I think this would do for now. The biggest issue is how it hogs the main thread while executing threads.

```
keypress-duration = <10>;
bindings =
    <&press LALT>,
    <&kp TAB>,
    <&sleep 30>.
    <&kp TAB>,
    <&release LALT>,
    <&kp LGUI(LEFT)>;
```

```
macro1 {
    keypress-duration = <30>;
    mode = "key-down";
    bindings = <&kp TAB> <&kp TAB>;
}
macro2 {
    keypress-duration = <10>;
    mode = "hold";
    bindings = <&kp LALT> <&macro1>;
}
macro 3 {
    keypress-duration = <10>;
    mode = "key-up";
    bindings = <&macro2> <&kp LGUI(LEFT)>;
}

bindings = <&macro3>
```

Reply:
> Seems like a default duration and <&press KC> <&sleep X> <&release KC> for a keypress with a different duration would be much simpler:

If we want to support arbitrary behaviors in macros, every behavior will need to be sent both a press and a release, and there will have to be a delay between them for at least some types of behaviors so key presses actually register. But then to do what you suggest, we'd need to have special cases for &press and &release so they *don't* have that delay.

I think we either need a system where you can manually control the duration of each item in the macro so that the macro just has to press, wait, and release for every kind of behavior no matter what it is, or we need a system where ZMK can change macros execute each type of behavior (without a bunch of "if behavior is X, do Y" blocks everywhere, because special cases like that tend to result in code that is difficult to maintain).

Reply:
Also note that using a "wait" behavior starts tying us to macros blocking anything else from running on the main thread. If all delays are part of the macros system, then we could fairly easily change macros to use delayed work items to allow other things to run. If we start relying on a behavior that sleeps the main thread, then making that switch would be much harder.

Reply:
If we changed the delay in `macro2` in my example to be 0, would that have the behavior you're looking for?

Reply:
> Manually controlling the delays between items could be especially useful in macros for games. For example:
> 
> ```
> // Automate something I'm too bad at this game to do.
> sleep = 0,
> bindings = 
>     <&down DOWN>, <&wait 17>,
>     <&down RIGHT>, <&wait 17>,
>     <&up DOWN>, <&wait 17>,
>     <&up RIGHT>, <&down Z>, <&wait 33>,
>     <&up Z>,  <&wait 134>,
>     <&down X>, <&down C>, <&wait 33>,
>     <&up X>, <&up C>;
> ```

I think macros in this sense may be more useful if we were to drop the extra behaviors and focus on emulating a human typing. Something more detached from the rest of the behaviors. This would give the macros the most flexibility, and not introduce behaviors that may not be needed (wait/sleep, key down, and key up).

```
gaming-macro? {
    macro = <KEYDOWN DOWN>, <WAIT 17>,
        <KEYDOWN RIGHT>, <WAIT 17>,
        <KEYUP DOWN>, <WAIT 17>,
        <KEYUP RIGHT>, <KEYDOWN Z>, <WAIT 33>,
        <KEYUP Z>,  <WAIT 134>,
        <KEYDOWN X>, <KEYDOWN C>, <WAIT 33>,
        <KEYUP X>, <KEYUP C>;
}
```

Reply:
I don't see an issue with making a special case for &press and &release. &wait could be interpreted by the macro code to introduce a delayed work item, it doesn't have to sleep and block the main thread.

Devicetree doesn't allow nesting like `&one &two`; you can't give references as arguments. A `&for` as you suggested could be created (the syntax would even work), but it would have to be interpreted by the macro to override the next `keypress-duration`.

Reply:
> Devicetree doesn't allow nesting like `&one &two`; you can't give references as arguments.

Macros are parameterless behaviors. It was tested here: <https://github.com/BrainWart/zmk-kyria-keymap/blob/simple-macro/config/kyria.keymap>

Reply:
@BrainWart I like the idea. I think we can achieve the same without having to resort to a special keymp syntax just for macros:

```
gaming-macro? {
    macro = <&press DOWN>, <&wait 17>,
        <&press RIGHT>, <&wait 17>,
        <&release DOWN>, <&wait 17>,
        <&release RIGHT>, <&press Z>, <&wait 33>,
        <&release Z>,  <&wait 134>,
        <&press X>, <&press C>, <&wait 33>,
        <&release X>, <&release C>;
}
```

This could work just fine.

Reply:
If I correctly understand that you mean `&press`, `&release`, and `&wait` would be special things interpreted only by macros and not regular behaviors that you could also stick into a keymap, I think that would work well. You could also C macro your ZMK macros to get something like `&kp`:

```c
#define MACRO_KP(key, duration) <&press (key)>, <&wait (duration)>, <&release (key)>
```

That does mean that you wouldn't be able to do things like put an RGB underglow command in the middle of a macro though, but I'm not sure how useful that would be.

The thing I mainly didn't like about the previous proposal was that it was taking regular keymap behaviors but then needing to layer special macro things on top of every behavior. e.g. `&kp` should be pressed, wait for X milliseconds, then released, `&press` and `&release` should just be pressed (and maybe immediately released?) and continue with no delay, `&wait` isn't pressed at all, and I have no idea what should happen if you use a `&mt`...

I vote for either:
1. Macros use their own set of commands that are independent from keymap behaviors.
2. Macros use keymap behaviors, but we always execute each behavior the same way (press, wait, release) and have some means to control how long each one is held.

Reply:
One of the strenghts of ZMK is the flexible keymap system using behaviors. Whatever we do with combos, we should support any and all behaviors from combos. This means option 1 (macros use their own set of commands) is a no-go in my opinion.

In option 2 you state that 'we always execute each behavior the same way (press, wait, release)'. I think we can make that work with an additional rule: wait is only executed when a behavior defines both the 'press' and 'release' bindings. Consider `&press` which is a behavior that only has a "press" binding, and `&release`, a behavior that only defines a "release" binding.  This allows the user to fully in control timing when using `&press` and `&release`, and also any new (user-defined) behaviors in the future.

The `&wait` must be a little special, when the macro encounters a `&wait` it should not call this behavior but only execute a "wait" with the specified delay. 

to summarize, I think we should:
* leverage the power of the devicetree in macros, which lets us use all existing behaviors
* add behaviors like `&press` and `&release` which could work inside and outside of macros
* add a behavior like `&wait` which is a no-op in a normal keymap but introduces a wait inside a macro